src/endpoint/s3/ops/s3_post_object_select.js

```diff
@@ -0,0 +1,86 @@
+/* Copyright (C) 2016 NooBaa */
+'use strict';
+
+const dbg = require('../../../util/debug_module')(__filename);
+const S3Error = require('../s3_errors').S3Error;
+const s3_utils = require('../s3_utils');
+const s3_select_utils = require('../../../util/s3select');
+const nb_native = require('../../../util/nb_native');
+const stream_utils = require('../../../util/stream_utils');
+
+/**
+ * https://docs.aws.amazon.com/AmazonS3/latest/API/API_SelectObjectContent.html
+ */
+async function post_object_select(req, res) {
+
+    if (!nb_native().S3Select) {
+        throw new S3Error(S3Error.S3SelectNotCompiled);
+    }
+
+    req.object_sdk.setup_abort_controller(req, res);
+    const agent_header = req.headers['user-agent'];
+    const noobaa_trigger_agent = agent_header && agent_header.includes('exec-env/NOOBAA_FUNCTION');
+    const encryption = s3_utils.parse_encryption(req);
+
+    const md_params = {
+        bucket: req.params.bucket,
+        key: req.params.key,
+        version_id: req.query.versionId,
+        encryption,
+    };
+    const object_md = await req.object_sdk.read_object_md(md_params);
+
+    const params = {
+        object_md,
+        obj_id: object_md.obj_id,
+        bucket: req.params.bucket,
+        key: req.params.key,
+        content_type: object_md.content_type,
+        noobaa_trigger_agent,
+        encryption,
+    };
+
+    //prepare s3select stream
+    const http_req_select_params = req.body.SelectObjectContentRequest;
+    const input_serialization = http_req_select_params.InputSerialization[0];
```

Comment:
what about OutputSerialization ?

Reply:
Some request parameters seem to be ignored.
Any reason why we cannot support it?

- [RequestProgress](https://docs.aws.amazon.com/AmazonS3/latest/API/API_SelectObjectContent.html#AmazonS3-SelectObjectContent-request-RequestProgress)
- [ScanRange](https://docs.aws.amazon.com/AmazonS3/latest/API/API_SelectObjectContent.html#AmazonS3-SelectObjectContent-request-ScanRange)
- [CompressionType](https://docs.aws.amazon.com/AmazonS3/latest/API/API_InputSerialization.html#AmazonS3-Type-InputSerialization-CompressionType)
- [OutputSerialization](https://docs.aws.amazon.com/AmazonS3/latest/API/API_SelectObjectContent.html#AmazonS3-SelectObjectContent-request-OutputSerialization)

Reply:
Generally, this are not implemeted (yet?). I can return 501 if one of them is in the request. Do you think it warrants a mail/meeting?

* Request Progress is optional. Also seems to somewhat contradict https://docs.aws.amazon.com/AmazonS3/latest/API/RESTSelectObjectAppendix.html which states "Stats message: Amazon S3 sends this message at the end of the request. It contains statistics about the query." I think this might be one of the differences between v1 and v2 of select object content.
* ScanRange - I can add it, if needed.
* CompressionType - I guess this is only relevant for FSNS?
* OutputSerlization - Implementing this sounds problematic because I don't see support for it in the s3select library so either we add ourselves to the lib or manipulate its output.

Reply:
- ScanRange sounds simple.
- CompressionType is unrelated to nsfs, it's for any object which is a compressed json/csv. We can defer that for now and return error if requested.
- OutputSerialization - I am not following how this is working then, what is the "default" output serialization format that the library responds with?

Reply:
*OutputSerialization - s3select response format is according to input - csv for csv, json for json. This is because it has different implementations for csv and json (s3selectEngine::json_object and s3selectEngine::csv_object).

Reply:
I see, so perhaps just validate that the request does not ask for converting to a different type, and accept if either not provided or matching the input type.